app/gameengine.py

```diff
@@ -2808,66 +2836,77 @@ def do_effect(self, effect_player : PlayerState, effect):
+                        case "self":
+                            source_card, _, _ = effect_player.find_card(effect["source_card_id"])
+                            target_holomems.append(source_card)
+                    cheer_options = []
+                    for holomem in target_holomems:
+                        if required_colors:
+                            matched_cheer = []
+                            for cheer in holomem["attached_cheer"]:
+                                if any(color in cheer["colors"] for color in required_colors):
+                                    matched_cheer.append(cheer)
+                            cheer_options += ids_from_cards(matched_cheer)
+                        else:
+                            cheer_options += ids_from_cards(holomem["attached_cheer"])
+                    after_archive_check_effect = {
+                        "player_id": effect_player_id,
+                        "effect_type": EffectType.EffectType_AfterArchiveCheerCheck,
                         "effect_player_id": effect_player_id,
-                        "all_card_seen": cheer_options,
-                        "cards_can_choose": cheer_options,
-                        "from_zone": "holomem",
-                        "to_zone": "archive",
-                        "amount_min": amount,
-                        "amount_max": amount,
-                        "reveal_chosen": True,
-                        "remaining_cards_action": "nothing",
+                        "previous_archive_count": len(effect_player.archive),
+                        "ability_source": ability_source
                     }
-                    self.broadcast_event(choose_event)
-                    self.set_decision({
-                        "decision_type": DecisionType.DecisionEffect_ChooseCardsForEffect,
-                        "decision_player": effect_player_id,
-                        "all_card_seen": cheer_options,
-                        "cards_can_choose": cheer_options,
-                        "from_zone": "holomem",
-                        "to_zone": "archive",
-                        "amount_min": amount,
-                        "amount_max": amount,
-                        "reveal_chosen": True,
-                        "remaining_cards_action": "nothing",
-                        "source_card_id": effect["source_card_id"],
-                        "effect_resolution": self.handle_choose_cards_result,
-                        "continuation": self.continue_resolving_effects,
-                    })
+                    self.add_effects_to_front([after_archive_check_effect])
+                    if amount == 0:
+                        self.continue_resolving_effects()
+                    elif amount == len(cheer_options):
+                        # Do it immediately.
+                        effect_player.archive_attached_cards(cheer_options)
+                        self.continue_resolving_effects()
```

Comment:
@daniel-k-taylor  How would you implement the ``before_archive_cheer`` timing that I added?
It works fine but I saw that it raised an exception that I didn't notice since it failed quietly and everything ran normally. I'm still kinda lost with the event handling so I'm not sure where to properly approach the problem.

Easy fix is to add a sanity check for ``effect_resolution_state`` in ``continue_resolving_effects`` that it is not ``None`` but that looks like a code smell to me.

Reply:
The changes in ``EffectType_ArchiveCheerFromHolomem`` are basically:
1. collect the ``before_archive_cheer`` effects
2. wrap the previous implem under a function that will act as the continuation.
3. resolve the before effects then do the continuation as normal.

i added ``self.continue_resolving_effects()`` for both instances where no choices are needed. (0 cards to archive and amount = target)

Reply:
I assume you modeled this after ArchiveFromHand/before_archive that was used for Lui and archiving cards.

I think what you're missing is that you need to set passed_on_continuation = True
This way the calling function knows not to keep trying to continue resolving effects